fairscale/experimental/nn/mevo.py

```diff
@@ -0,0 +1,483 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+
+from typing import Any, Tuple
+
+import torch
+from torch import nn
+import torch.distributed as dist
+import torch.nn.functional as F
+
+# Debugging flag to enable some prints. Useful to debug with FSDP.
+DEBUG = False
+
+
+def _next_power_of_2_or_max(n: int, max_n: int) -> int:
+    """ Return the smallest power of 2 greater than or equal to n, with a limit.
+
+        Useful when used in splitting a tensor into chunks with power-of-2 sizes.
+    """
+    # special case, just split to 1 element chunks.
+    if n == 0:
+        return 1
+    orig_n = n
+    n -= 1
+    n |= n >> 1
+    n |= n >> 2
+    n |= n >> 4
+    n |= n >> 8
+    n |= n >> 16
+    n += 1
+    assert n >= orig_n, f"{n} vs. {orig_n}"
+    assert bin(n).count("1") == 1, bin(n)  # Catch the case n is too large for this function.
+    if n > max_n:
+        return max_n
+    return n
+
+
+def _reshape_inputs(input: torch.Tensor, target: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+    """Convert 3D inputs to 2D for this kernel"""
+    if len(input.shape) == 3:
+        input = input.reshape(-1, input.shape[2])
+    if len(target.shape) == 2:
+        target = target.reshape(-1)
```

Comment:
will this be a 1-D or 2D target?

Reply:
target is always 1D in this layer. We just reshape the input when the input is 2D. input is also always reshaped from 3d to 2d.